homeassistant/components/abode/config_flow.py

```diff
@@ -26,48 +27,151 @@ def __init__(self):
+            )
+
+        except (AbodeException, ConnectTimeout, HTTPError) as ex:
+            if ex.errcode == MFA_CODE_REQUIRED[0]:
+                return await self.async_step_mfa()
+
+            LOGGER.error("Unable to connect to Abode: %s", ex)
+
+            if ex.errcode == HTTP_BAD_REQUEST:
+                errors = {"base": "invalid_auth"}
+            else:
+                errors = {"base": "cannot_connect"}
+
+        if errors:
+            return self.async_show_form(
+                step_id=step_id, data_schema=vol.Schema(self.data_schema), errors=errors
+            )
+
+        return await self._async_create_entry(
+            {
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_POLLING: self._polling,
+            }
+        )
+
+    async def _async_create_entry(self, user_input):
+        """Create the config entry."""
+        existing_entry = await self.async_set_unique_id(self._username)
+
+        if existing_entry:
+            self.hass.config_entries.async_update_entry(existing_entry, data=user_input)
+            # Force reload the Abode config entry otherwise devices will remain unavailable
+            self.hass.async_create_task(
+                self.hass.config_entries.async_reload(existing_entry.entry_id)
+            )
+
+            return self.async_abort(reason="reauth_successful")
+
+        return self.async_create_entry(title=self._username, data=user_input)
+
     async def async_step_user(self, user_input=None):
         """Handle a flow initialized by the user."""
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
 
-        if not user_input:
-            return self._show_form()
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=vol.Schema(self.data_schema)
+            )
+
+        self._username = user_input[CONF_USERNAME]
+        self._password = user_input[CONF_PASSWORD]
+        self._polling = user_input.get(CONF_POLLING, False)
 
-        username = user_input[CONF_USERNAME]
-        password = user_input[CONF_PASSWORD]
-        polling = user_input.get(CONF_POLLING, False)
-        cache = self.hass.config.path(DEFAULT_CACHEDB)
+        return await self._async_abode_login(step_id="user")
+
+    async def async_step_mfa(self, user_input=None):
+        """Handle multi-factor authentication (MFA) with Abode."""
+        errors = {}
+
+        if user_input is None:
+            return self.async_show_form(
+                step_id="mfa", data_schema=vol.Schema(self.mfa_data_schema)
+            )
 
         try:
+            mfa_code = user_input[CONF_MFA]
+            abode = Abode(
+                auto_login=False,
+                get_devices=False,
+                get_automations=False,
+                cache_path=self._cache,
+            )
```

Comment:
I'll move these out of the `try:` block since we're not trying to catch anything.